### PR TITLE
Clearing LinkCache on key change

### PR DESCRIPTION
**Pull request by sojanpr**
Clearing the branch link cache on a  new key Branch key

### CODE REVIEW

File: Branch-SDK/src/io/branch/referral/Branch.java [hunk 1]
```diff
@@ -393,7 +393,11 @@ public static Branch getInstance(Context context, String branchKey) {
 		}
 		branchReferral_.context_ = context.getApplicationContext();
 		if (branchKey.startsWith("key_")) {
-			branchReferral_.prefHelper_.setBranchKey(branchKey);
+			boolean isNewBranchKeySet = branchReferral_.prefHelper_.setBranchKey(branchKey);
+			//on setting a new key clear link cache
+			if (isNewBranchKeySet) {
+				branchReferral_.linkCache_.clear();
```
Comment (shortstuffsushi) [inline Branch-SDK/src/io/branch/referral/Branch.java]: Link cache doesn't get persisted between launches, does it?

Comment (shortstuffsushi) [inline Branch-SDK/src/io/branch/referral/Branch.java]: Instead, the goal of "clearing when key changes" is supposed to remove identity, install/session params, etc.

Comment (shortstuffsushi) [inline Branch-SDK/src/io/branch/referral/Branch.java]: See the [iOS SDK implementation](https://github.com/BranchMetrics/Branch-iOS-Deep-Linking-SDK/blob/aa0fa04364ddbc5abb19461e01e1f93b5e48e84e/Branch-SDK/Branch-SDK/Branch.m#L829)

Comment (sojanpr) [inline Branch-SDK/src/io/branch/referral/Branch.java]: Link Cache is not persisted across launches how ever there is option for user to set a new key in the same session by calling the API `Branch getInstance(Context context, String branchKey)`

Comment (shortstuffsushi) [inline Branch-SDK/src/io/branch/referral/Branch.java]: In that case, you should just do both :smiley:

**Comment by shortstuffsushi** [comment]
Hmm, the iOS SDK is not clearing the Link Cache (probably should be). However, I noticed it is clearing the queue, which is not being done here. Does that need to occur as well?

File: Branch-SDK/src/io/branch/referral/PrefHelper.java [hunk 2]
```diff
@@ -340,13 +340,20 @@ public String getAppKey() {
...
+	 * @param key A {@link String} representing Branch Key.
+	 * @return A {@link Boolean} which is true if the key set is a new key. On Setting a new key need to clear all preference items.
+	 */
+	public boolean setBranchKey(String key) {
 		Branch_Key = key;
 		String currentBranchKey = getString(KEY_BRANCH_KEY);
 		if (key == null || currentBranchKey == null || !currentBranchKey.equals(key)) {
 			clearPrefOnBranchKeyChange();
 			setString(KEY_BRANCH_KEY, key);
+			return true;
```
Comment (aaustin) [inline Branch-SDK/src/io/branch/referral/PrefHelper.java]: nice and clean

File: Branch-SDK/src/io/branch/referral/Branch.java (hunk 1 above)
Comment (aaustin) [inline Branch-SDK/src/io/branch/referral/Branch.java]: Agreed - need to clear more than link cache. all Branch ids.

**Comment by aaustin** [comment]
:+1:

**Comment by sojanpr** [comment]
Added changes to clear pending server request on key change.

**Comment by shortstuffsushi** [comment]
:+1:

**Pull request merged**